File: src/bigclaw/evaluation.py
from dataclasses import dataclass, field
from html import escape
from pathlib import Path
from typing import List, Optional

from .models import Task
from .observability import ObservabilityLedger
from .scheduler import ExecutionRecord, Scheduler
from .reports import write_report


@dataclass
class EvaluationCriterion:
    name: str
    weight: int
    passed: bool
    detail: str


@dataclass
class BenchmarkCase:
    case_id: str
    task: Task
    expected_medium: Optional[str] = None
    expected_approved: Optional[bool] = None
    expected_status: Optional[str] = None
    require_report: bool = False


@dataclass
class ReplayRecord:
    task: Task
    run_id: str
    medium: str
    approved: bool
    status: str

    @classmethod
    def from_execution(cls, task: Task, run_id: str, record: ExecutionRecord) -> "ReplayRecord":
        return cls(
            task=task,
            run_id=run_id,
            medium=record.decision.medium,
            approved=record.decision.approved,
            status=record.run.status,
        )


@dataclass
class ReplayOutcome:
    matched: bool
    replay_record: ReplayRecord
    mismatches: List[str] = field(default_factory=list)
    report_path: Optional[str] = None


@dataclass
class BenchmarkResult:
    case_id: str
    score: int
    passed: bool
    criteria: List[EvaluationCriterion]
    record: ExecutionRecord
    replay: ReplayOutcome
    detail_page_path: Optional[str] = None


@dataclass
class BenchmarkComparison:
    case_id: str
    baseline_score: int
    current_score: int
    delta: int
    changed: bool


@dataclass
class BenchmarkSuiteResult:
    results: List[BenchmarkResult]
    version: str = "current"

    @property
    def score(self) -> int:
        if not self.results:
            return 0
        return round(sum(result.score for result in self.results) / len(self.results))

    @property
    def passed(self) -> bool:
        return all(result.passed for result in self.results)

    def compare(self, baseline: "BenchmarkSuiteResult") -> List[BenchmarkComparison]:
        baseline_by_case = {result.case_id: result for result in baseline.results}
        comparisons = []
        for result in self.results:
            baseline_result = baseline_by_case.get(result.case_id)
            baseline_score = baseline_result.score if baseline_result else 0
            delta = result.score - baseline_score
            comparisons.append(
                BenchmarkComparison(
                    case_id=result.case_id,
                    baseline_score=baseline_score,
                    current_score=result.score,
                    delta=delta,
                    changed=delta != 0,
                )
            )
        return comparisons


class BenchmarkRunner:
    def __init__(self, scheduler: Optional[Scheduler] = None, storage_dir: Optional[str] = None):
        self.scheduler = scheduler or Scheduler()
        self.storage_dir = Path(storage_dir) if storage_dir else None

    def run_case(self, case: BenchmarkCase) -> BenchmarkResult:
        ledger = ObservabilityLedger(str(self._case_path(case.case_id, "ledger.json")))
        report_path = None
        if case.require_report:
            report_path = str(self._case_path(case.case_id, "task-run.md"))

        run_id = f"benchmark-{case.case_id}"
        record = self.scheduler.execute(
            case.task,
            run_id=run_id,
            ledger=ledger,
            report_path=report_path,
            actor="benchmark-runner",
        )
        criteria = self._evaluate(case, record)
        replay = self.replay(ReplayRecord.from_execution(case.task, run_id, record))
        total_weight = sum(item.weight for item in criteria)
        earned_weight = sum(item.weight for item in criteria if item.passed)
        score = round((earned_weight / total_weight) * 100) if total_weight else 0
        passed = all(item.passed for item in criteria) and replay.matched
        detail_page_path = None
        if self.storage_dir is not None:
            detail_page_path = str(self._case_path(case.case_id, "run-detail.html"))
            write_report(detail_page_path, render_run_replay_index_page(case.case_id, record, replay, criteria))
        return BenchmarkResult(
            case_id=case.case_id,
            score=score,
            passed=passed,
            criteria=criteria,
            record=record,
            replay=replay,
            detail_page_path=detail_page_path,
        )

    def run_suite(self, cases: List[BenchmarkCase], version: str = "current") -> BenchmarkSuiteResult:
        return BenchmarkSuiteResult(
            results=[self.run_case(case) for case in cases],
            version=version,
        )

    def replay(self, replay_record: ReplayRecord) -> ReplayOutcome:
        ledger = ObservabilityLedger(str(self._case_path(replay_record.run_id, "replay-ledger.json")))
        replayed = self.scheduler.execute(
            replay_record.task,
            run_id=f"{replay_record.run_id}-replay",
            ledger=ledger,
            actor="benchmark-replay",
        )
        observed = ReplayRecord.from_execution(
            replay_record.task,
            replay_record.run_id,
            replayed,
        )
        mismatches = []
        if observed.medium != replay_record.medium:
            mismatches.append(f"medium expected {replay_record.medium} got {observed.medium}")
        if observed.approved != replay_record.approved:
            mismatches.append(
                f"approved expected {replay_record.approved} got {observed.approved}"
            )
        if observed.status != replay_record.status:
            mismatches.append(f"status expected {replay_record.status} got {observed.status}")
        report_path = None
        if self.storage_dir is not None:
            report_path = str(self._case_path(replay_record.run_id, "replay.html"))
            write_report(report_path, render_replay_detail_page(replay_record, observed, mismatches))
        return ReplayOutcome(
            matched=not mismatches,
            replay_record=observed,
            mismatches=mismatches,
            report_path=report_path,
        )

    def _evaluate(self, case: BenchmarkCase, record: ExecutionRecord) -> List[EvaluationCriterion]:
        return [
            self._criterion(
                name="decision-medium",
                weight=40,
                expected=case.expected_medium,
                actual=record.decision.medium,
            ),
            self._criterion(
                name="approval-gate",
                weight=30,
                expected=case.expected_approved,
                actual=record.decision.approved,
            ),
            self._criterion(
                name="final-status",
                weight=20,
                expected=case.expected_status,
                actual=record.run.status,
            ),
            EvaluationCriterion(
                name="report-artifact",
                weight=10,
                passed=(not case.require_report) or bool(record.report_path),
                detail=(
                    "report emitted"
                    if (not case.require_report) or bool(record.report_path)
                    else "report missing"
                ),
            ),
        ]

    def _criterion(self, name: str, weight: int, expected: Optional[object], actual: object) -> EvaluationCriterion:
        if expected is None:
            return EvaluationCriterion(name=name, weight=weight, passed=True, detail="not asserted")
        passed = expected == actual
        detail = f"expected {expected} got {actual}"
        return EvaluationCriterion(name=name, weight=weight, passed=passed, detail=detail)

    def _case_path(self, case_id: str, file_name: str) -> Path:
        if self.storage_dir is None:
            return Path(file_name)
        return self.storage_dir / case_id / file_name


def render_benchmark_suite_report(
    suite: BenchmarkSuiteResult,
    baseline: Optional[BenchmarkSuiteResult] = None,
) -> str:
    lines = [
        "# Benchmark Suite Report",
        "",
        f"- Version: {suite.version}",
        f"- Cases: {len(suite.results)}",
        f"- Passed: {suite.passed}",
        f"- Score: {suite.score}",
        "",
        "## Cases",
        "",
    ]

    if suite.results:
        lines.extend(
            f"- {result.case_id}: score={result.score} passed={result.passed} replay={result.replay.matched}"
            for result in suite.results
        )
    else:
        lines.append("- None")

    lines.extend(["", "## Comparison", ""])
    if baseline is None:
        lines.append("- No baseline provided")
    else:
        lines.append(f"- Baseline Version: {baseline.version}")
        lines.append(f"- Score Delta: {suite.score - baseline.score}")
        comparisons = suite.compare(baseline)
        if comparisons:
            lines.extend(
                f"- {comparison.case_id}: baseline={comparison.baseline_score} current={comparison.current_score} delta={comparison.delta}"
                for comparison in comparisons
            )
        else:
            lines.append("- No comparable cases")

    return "\n".join(lines) + "\n"


def render_replay_detail_page(expected: ReplayRecord, observed: ReplayRecord, mismatches: List[str]) -> str:
    mismatch_items = "".join(f"<li>{escape(item)}</li>" for item in mismatches) or "<li>None</li>"
    return f"""<!doctype html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <title>Replay Detail · {escape(expected.run_id)}</title>
  <style>
    :root {{ color-scheme: light dark; font-family: -apple-system, BlinkMacSystemFont, \"Segoe UI\", sans-serif; }}
    body {{ margin: 2rem auto; max-width: 880px; padding: 0 1rem 3rem; line-height: 1.5; }}
    table {{ border-collapse: collapse; width: 100%; margin: 1rem 0 1.5rem; }}
    th, td {{ border: 1px solid #cbd5e1; padding: 0.6rem; text-align: left; }}
    th {{ background: rgba(148, 163, 184, 0.12); }}
  </style>
</head>
<body>
  <h1>Replay Detail</h1>
  <p>Task <strong>{escape(expected.task.task_id)}</strong> · baseline run <code>{escape(expected.run_id)}</code></p>
  <table>
    <thead>
      <tr><th>Field</th><th>Expected</th><th>Observed</th></tr>
    </thead>
    <tbody>
      <tr><td>Medium</td><td>{escape(expected.medium)}</td><td>{escape(observed.medium)}</td></tr>
      <tr><td>Approved</td><td>{escape(str(expected.approved))}</td><td>{escape(str(observed.approved))}</td></tr>
      <tr><td>Status</td><td>{escape(expected.status)}</td><td>{escape(observed.status)}</td></tr>
    </tbody>
  </table>
  <h2>Mismatches</h2>
  <ul>{mismatch_items}</ul>
</body>
</html>
"""


def render_run_replay_index_page(
    case_id: str,
    record: ExecutionRecord,
    replay: ReplayOutcome,
    criteria: List[EvaluationCriterion],
) -> str:
    report_path = escape(record.report_path or "n/a")
    detail_path = escape(str(Path(record.report_path).with_suffix(".html"))) if record.report_path else "n/a"
    replay_path = escape(replay.report_path or "n/a")
    criteria_items = "".join(
        f"<li><strong>{escape(item.name)}</strong>: {escape(item.detail)} · passed={escape(str(item.passed))}</li>"
        for item in criteria
    ) or "<li>None</li>"
    mismatch_items = "".join(f"<li>{escape(item)}</li>" for item in replay.mismatches) or "<li>None</li>"
    return f"""<!doctype html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <title>Run Detail Index · {escape(case_id)}</title>
  <style>
    :root {{ color-scheme: light dark; font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", sans-serif; }}
    body {{ margin: 2rem auto; max-width: 900px; padding: 0 1rem 3rem; line-height: 1.5; }}
    .card {{ border: 1px solid #cbd5e1; border-radius: 10px; padding: 1rem; margin: 1rem 0; background: rgba(148, 163, 184, 0.08); }}
    code {{ font-size: 0.95em; }}
  </style>
</head>
<body>
  <h1>Run Detail Index</h1>
  <p>Benchmark case <strong>{escape(case_id)}</strong> · task <strong>{escape(record.run.task_id)}</strong></p>
  <div class="card">
    <h2>Execution</h2>
    <p>Status: <strong>{escape(record.run.status)}</strong> · Medium: <strong>{escape(record.decision.medium)}</strong></p>
    <ul>
      <li>Markdown report: <code>{report_path}</code></li>
      <li>Run detail page: <code>{detail_path}</code></li>
      <li>Replay page: <code>{replay_path}</code></li>
    </ul>
  </div>
  <div class="card">
    <h2>Acceptance Criteria</h2>
    <ul>{criteria_items}</ul>
  </div>
  <div class="card">
    <h2>Replay Mismatches</h2>
    <ul>{mismatch_items}</ul>
  </div>
</body>
</html>
"""

File: bigclaw-go/internal/evaluation/evaluation.go
package evaluation

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"bigclaw-go/internal/domain"
	"bigclaw-go/internal/scheduler"
)

type Criterion struct {
	Name   string
	Weight int
	Passed bool
	Detail string
}

type BenchmarkCase struct {
	CaseID           string
	Task             domain.Task
	ExpectedExecutor domain.ExecutorKind
	ExpectAccepted   *bool
	ExpectedStatus   string
	RequireReport    bool
}

type ReplayRecord struct {
	Task     domain.Task
	RunID    string
	Executor domain.ExecutorKind
	Accepted bool
	Status   string
}

type ReplayOutcome struct {
	Matched      bool
	ReplayRecord ReplayRecord
	Mismatches   []string
	ReportPath   string
}

type BenchmarkResult struct {
	CaseID         string
	Score          int
	Passed         bool
	Criteria       []Criterion
	Replay         ReplayOutcome
	ReportPath     string
	DetailPagePath string
}

type BenchmarkComparison struct {
	CaseID        string
	BaselineScore int
	CurrentScore  int
	Delta         int
	Changed       bool
}

type BenchmarkSuiteResult struct {
	Results []BenchmarkResult
	Version string
}

func (suite BenchmarkSuiteResult) Score() int {
	if len(suite.Results) == 0 {
		return 0
	}
	total := 0
	for _, result := range suite.Results {
		total += result.Score
	}
	return int(float64(total)/float64(len(suite.Results)) + 0.5)
}

func (suite BenchmarkSuiteResult) Passed() bool {
	for _, result := range suite.Results {
		if !result.Passed {
			return false
		}
	}
	return true
}

func (suite BenchmarkSuiteResult) Compare(baseline BenchmarkSuiteResult) []BenchmarkComparison {
	byCase := map[string]BenchmarkResult{}
	for _, result := range baseline.Results {
		byCase[result.CaseID] = result
	}
	comparisons := make([]BenchmarkComparison, 0, len(suite.Results))
	for _, result := range suite.Results {
		baselineScore := 0
		if existing, ok := byCase[result.CaseID]; ok {
			baselineScore = existing.Score
		}
		delta := result.Score - baselineScore
		comparisons = append(comparisons, BenchmarkComparison{
			CaseID:        result.CaseID,
			BaselineScore: baselineScore,
			CurrentScore:  result.Score,
			Delta:         delta,
			Changed:       delta != 0,
		})
	}
	return comparisons
}

type Runner struct {
	Scheduler  *scheduler.Scheduler
	StorageDir string
}

func (runner Runner) RunCase(tc BenchmarkCase) (BenchmarkResult, error) {
	runtime := runner.scheduler()
	runID := "benchmark-" + tc.CaseID
	decision := runtime.Decide(tc.Task, scheduler.QuotaSnapshot{})
	record := ReplayRecord{
		Task:     tc.Task,
		RunID:    runID,
		Executor: decision.Assignment.Executor,
		Accepted: decision.Accepted,
		Status:   decisionStatus(decision.Accepted),
	}

	reportPath := ""
	if tc.RequireReport {
		reportPath = runner.casePath(tc.CaseID, "task-run.md")
		if err := writeText(reportPath, renderTaskRunMarkdown(record)); err != nil {
			return BenchmarkResult{}, err
		}
	}

	criteria := evaluateCriteria(tc, record, reportPath)
	replay, err := runner.Replay(record)
	if err != nil {
		return BenchmarkResult{}, err
	}
	score := scoreCriteria(criteria)
	passed := replay.Matched
	for _, criterion := range criteria {
		if !criterion.Passed {
			passed = false
			break
		}
	}

	detailPath := ""
	if runner.StorageDir != "" {
		detailPath = runner.casePath(tc.CaseID, "run-detail.html")
		if err := writeText(detailPath, RenderRunReplayIndexPage(tc.CaseID, record, replay, criteria, reportPath)); err != nil {
			return BenchmarkResult{}, err
		}
	}

	return BenchmarkResult{
		CaseID:         tc.CaseID,
		Score:          score,
		Passed:         passed,
		Criteria:       criteria,
		Replay:         replay,
		ReportPath:     reportPath,
		DetailPagePath: detailPath,
	}, nil
}

func (runner Runner) RunSuite(cases []BenchmarkCase, version string) (BenchmarkSuiteResult, error) {
	results := make([]BenchmarkResult, 0, len(cases))
	for _, tc := range cases {
		result, err := runner.RunCase(tc)
		if err != nil {
			return BenchmarkSuiteResult{}, err
		}
		results = append(results, result)
	}
	return BenchmarkSuiteResult{Results: results, Version: version}, nil
}

func (runner Runner) Replay(record ReplayRecord) (ReplayOutcome, error) {
	runtime := runner.scheduler()
	decision := runtime.Decide(record.Task, scheduler.QuotaSnapshot{})
	observed := ReplayRecord{
		Task:     record.Task,
		RunID:    record.RunID,
		Executor: decision.Assignment.Executor,
		Accepted: decision.Accepted,
		Status:   decisionStatus(decision.Accepted),
	}

	mismatches := []string{}
	if observed.Executor != record.Executor {
		mismatches = append(mismatches, fmt.Sprintf("executor expected %s got %s", record.Executor, observed.Executor))
	}
	if observed.Accepted != record.Accepted {
		mismatches = append(mismatches, fmt.Sprintf("accepted expected %t got %t", record.Accepted, observed.Accepted))
	}
	if observed.Status != record.Status {
		mismatches = append(mismatches, fmt.Sprintf("status expected %s got %s", record.Status, observed.Status))
	}

	reportPath := ""
	if runner.StorageDir != "" {
		reportPath = runner.casePath(record.RunID, "replay.html")
		if err := writeText(reportPath, RenderReplayDetailPage(record, observed, mismatches)); err != nil {
			return ReplayOutcome{}, err
		}
	}

	return ReplayOutcome{
		Matched:      len(mismatches) == 0,
		ReplayRecord: observed,
		Mismatches:   mismatches,
		ReportPath:   reportPath,
	}, nil
}

func RenderBenchmarkSuiteReport(suite BenchmarkSuiteResult, baseline *BenchmarkSuiteResult) string {
	lines := []string{
		"# Benchmark Suite Report",
		"",
		fmt.Sprintf("- Version: %s", suite.Version),
		fmt.Sprintf("- Cases: %d", len(suite.Results)),
		fmt.Sprintf("- Passed: %t", suite.Passed()),
		fmt.Sprintf("- Score: %d", suite.Score()),
		"",
		"## Cases",
		"",
	}
	if len(suite.Results) == 0 {
		lines = append(lines, "- None")
	} else {
		for _, result := range suite.Results {
			lines = append(lines, fmt.Sprintf("- %s: score=%d passed=%t replay=%t", result.CaseID, result.Score, result.Passed, result.Replay.Matched))
		}
	}
	lines = append(lines, "", "## Comparison", "")
	if baseline == nil {
		lines = append(lines, "- No baseline provided")
	} else {
		lines = append(lines, fmt.Sprintf("- Baseline Version: %s", baseline.Version))
		lines = append(lines, fmt.Sprintf("- Score Delta: %d", suite.Score()-baseline.Score()))
		for _, comparison := range suite.Compare(*baseline) {
			lines = append(lines, fmt.Sprintf("- %s: baseline=%d current=%d delta=%d", comparison.CaseID, comparison.BaselineScore, comparison.CurrentScore, comparison.Delta))
		}
	}
	return strings.Join(lines, "\n") + "\n"
}

func RenderReplayDetailPage(expected, observed ReplayRecord, mismatches []string) string {
	lines := []string{
		"<html><body>",
		"<h1>Replay Detail</h1>",
		"<h2>Timeline / Log Sync</h2>",
		"<h2>Split View</h2>",
		"<h2>Reports</h2>",
		fmt.Sprintf("<p>expected executor=%s observed executor=%s</p>", expected.Executor, observed.Executor),
		fmt.Sprintf("<p>expected status=%s observed status=%s</p>", expected.Status, observed.Status),
	}
	if len(mismatches) == 0 {
		lines = append(lines, "<p>None</p>")
	} else {
		for _, mismatch := range mismatches {
			lines = append(lines, "<p>"+mismatch+"</p>")
		}
	}
	lines = append(lines, "</body></html>")
	return strings.Join(lines, "\n")
}

func RenderRunReplayIndexPage(caseID string, record ReplayRecord, replay ReplayOutcome, criteria []Criterion, reportPath string) string {
	replayPath := replay.ReportPath
	if replayPath == "" {
		replayPath = "n/a"
	}
	if reportPath == "" {
		reportPath = "n/a"
	}
	detailPath := "n/a"
	if reportPath != "n/a" {
		detailPath = strings.TrimSuffix(reportPath, filepath.Ext(reportPath)) + ".html"
	}
	lines := []string{
		"<html><body>",
		"<h1>Run Detail Index</h1>",
		"<h2>Timeline / Log Sync</h2>",
		"<h2>Acceptance</h2>",
		"<h2>Replay</h2>",
		"<h2>Reports</h2>",
		fmt.Sprintf("<p>case=%s task=%s executor=%s status=%s</p>", caseID, record.Task.ID, record.Executor, record.Status),
		fmt.Sprintf("<p>%s</p>", reportPath),
		fmt.Sprintf("<p>%s</p>", detailPath),
		fmt.Sprintf("<p>%s</p>", replayPath),
	}
	for _, criterion := range criteria {
		lines = append(lines, fmt.Sprintf("<p>%s: %s</p>", criterion.Name, criterion.Detail))
	}
	lines = append(lines, "</body></html>")
	return strings.Join(lines, "\n")
}

func evaluateCriteria(tc BenchmarkCase, record ReplayRecord, reportPath string) []Criterion {
	criteria := []Criterion{
		criterionExecutor(tc.ExpectedExecutor, record.Executor),
		criterionAccepted(tc.ExpectAccepted, record.Accepted),
		criterionStatus(tc.ExpectedStatus, record.Status),
	}
	criteria = append(criteria, Criterion{
		Name:   "report-artifact",
		Weight: 10,
		Passed: !tc.RequireReport || reportPath != "",
		Detail: ternary(!tc.RequireReport || reportPath != "", "report emitted", "report missing"),
	})
	return criteria
}

func criterionExecutor(expected, actual domain.ExecutorKind) Criterion {
	if expected == "" {
		return Criterion{Name: "decision-executor", Weight: 40, Passed: true, Detail: "not asserted"}
	}
	return Criterion{Name: "decision-executor", Weight: 40, Passed: expected == actual, Detail: fmt.Sprintf("expected %s got %s", expected, actual)}
}

func criterionAccepted(expected *bool, actual bool) Criterion {
	if expected == nil {
		return Criterion{Name: "approval-gate", Weight: 30, Passed: true, Detail: "not asserted"}
	}
	return Criterion{Name: "approval-gate", Weight: 30, Passed: *expected == actual, Detail: fmt.Sprintf("expected %t got %t", *expected, actual)}
}

func criterionStatus(expected, actual string) Criterion {
	if expected == "" {
		return Criterion{Name: "final-status", Weight: 20, Passed: true, Detail: "not asserted"}
	}
	return Criterion{Name: "final-status", Weight: 20, Passed: expected == actual, Detail: fmt.Sprintf("expected %s got %s", expected, actual)}
}

func scoreCriteria(criteria []Criterion) int {
	totalWeight := 0
	earned := 0
	for _, item := range criteria {
		totalWeight += item.Weight
		if item.Passed {
			earned += item.Weight
		}
	}
	if totalWeight == 0 {
		return 0
	}
	return int(float64(earned)/float64(totalWeight)*100 + 0.5)
}

func decisionStatus(accepted bool) string {
	if accepted {
		return "approved"
	}
	return "needs-approval"
}

func (runner Runner) scheduler() *scheduler.Scheduler {
	if runner.Scheduler != nil {
		return runner.Scheduler
	}
	return scheduler.New()
}

func (runner Runner) casePath(caseID, name string) string {
	if runner.StorageDir == "" {
		return name
	}
	return filepath.Join(runner.StorageDir, caseID, name)
}

func renderTaskRunMarkdown(record ReplayRecord) string {
	return fmt.Sprintf("# Task Run\n\n- Task: %s\n- Executor: %s\n- Status: %s\n", record.Task.ID, record.Executor, record.Status)
}

func writeText(path, body string) error {
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return err
	}
	return os.WriteFile(path, []byte(body), 0o644)
}

func ternary[T any](condition bool, yes, no T) T {
	if condition {
		return yes
	}
	return no
}
